### ISSUE TITLE
PAML suggestion: calculation of time / energy cost for a given protocol.

### ISSUE THREAD

**Issue by photocyte** [Bioprotocols/paml#144]
Hi there folks,

I'm using this PAML issues section as a general feedback/suggestion box. Not necessarily saying this is a pressing issue nor related to the code in this repo itself. Haven't searched if this is already addressed. 

A "killer app" for PAML could be making the cost (in terms of reagent cost) and time spent (i.e. instrument time, worker time) for a given protocol, more explicitly calculated & editable/"experimentable". I.e., what is the impact if I include an additional replicate in my protocol (or, subsection of a protocol), or if I miniaturize the volumes involved in a given step ("assay miniaturization" minimizing reagent cost). Am I actually mislead that the cost of some small unit operation that I have to do many times, adds up to the dominant cost of a protocol? If I want to optimize the cost or turnaround time of a given protocol, by sending off samples to some off-site core facility or cloud lab for a given specialized step, how does that tradeoff factor into the rest of the protocol?

This would be in the context of, it's hard / laborious to manually do this up-front for a complex hybrid manual/automated protocol, other than an intuitive feel from being an experienced lab worker.

**Comment by danbryce** [comment]
Yes, this is the direction we were originally going with protocol languages at the start of SD2.  We have the elements needed to do time projections, but would need something new to model costs.  

Modeling cost is tricky in some cases because its not just the cost of a primitive, but also the size of the data it operates on.  For example, a transfer operation uses more time or cost to transfer several samples over just one.

**Comment by rpgoldman** [comment]
Related to the point about using this as a location for discussion, maybe we should enable the "Discussions" feature on this GitHub repo?

I don't mind using issues for the purpose as such, but it makes it harder to triage and to close issues if they don't have a resolution.  The Discussions feature, a lightweight forum, would let us keep this straight. And you can transform issues into discussions, too. @danbryce @jakebeal
